### PR TITLE
[blockdao] pass ChainID to checkIndexers()

### PR DESCRIPTION
**Pull request by dustinxie**

### CODE REVIEW

File: blockchain/blockdao/blockdao.go [hunk 1]
```diff
@@ -105,37 +105,40 @@ func (dao *blockDAO) Start(ctx context.Context) error {
...
-func (dao *blockDAO) fillWithBlockInfoAsTip(ctx context.Context, height uint64) (context.Context, error) {
-	g, ok := genesis.ExtractGenesisContext(ctx)
-	if !ok {
-		return nil, errors.New("failed to find blockchain ctx")
+func (dao *blockDAO) fillWithBlockInfoAsTip(ctx context.Context, chainID uint32, height uint64) (context.Context, error) {
+	bcCtx := protocol.BlockchainCtx{
+		Tip: protocol.TipInfo{
+			Height: height,
+		},
+		ChainID: chainID,
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:113]: found this during testing
`ChainID` is missing in the returned context

File: blockchain/blockdao/blockdao.go [hunk 2]
```diff
@@ -105,37 +105,40 @@ func (dao *blockDAO) Start(ctx context.Context) error {
...
 	}
-	bcCtx := protocol.BlockchainCtx{}
 	if height == 0 {
-		bcCtx.Tip = protocol.TipInfo{
-			Height:    0,
-			Hash:      g.Hash(),
-			Timestamp: time.Unix(g.Timestamp, 0),
+		g, ok := genesis.ExtractGenesisContext(ctx)
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:116]: g is only used in here, so move in here

File: e2etest/local_test.go [hunk 3]
```diff
@@ -575,17 +572,21 @@ func TestStartExistingBlockchain(t *testing.T) {
...
 	// Recover to height 3 from empty state DB
 	cfg.DB.DbPath = cfg.Chain.ChainDBPath
 	cfg.DB.CompressLegacy = cfg.Chain.CompressBlock
-	dao = blockdao.NewBlockDAO(nil, cfg.DB)
-	require.NoError(dao.Start(genesis.WithGenesisContext(ctx, cfg.Genesis)))
+	dao := blockdao.NewBlockDAO(nil, cfg.DB)
+	require.NoError(dao.Start(protocol.WithBlockchainCtx(
+		genesis.WithGenesisContext(ctx, cfg.Genesis),
+		protocol.BlockchainCtx{
+			ChainID: cfg.Chain.ID,
```
Comment (dustinxie) [inline e2etest/local_test.go:588]: pass `ChainID` to `dao.Start()`

File: blockchain/blockdao/blockdao.go [hunk 4]
```diff
@@ -105,37 +105,40 @@ func (dao *blockDAO) Start(ctx context.Context) error {
...
-	if !ok {
-		return nil, errors.New("failed to find blockchain ctx")
+func (dao *blockDAO) fillWithBlockInfoAsTip(ctx context.Context, chainID uint32, height uint64) (context.Context, error) {
+	bcCtx := protocol.BlockchainCtx{
+		Tip: protocol.TipInfo{
+			Height: height,
+		},
+		ChainID: chainID,
 	}
-	bcCtx := protocol.BlockchainCtx{}
```
Comment (CoderZhi) [inline blockchain/blockdao/blockdao.go:113]: `bcCtx := protocol.GetBlockchainCtx(ctx)`
is this the only line we need to change?

Comment (dustinxie) [reply blockchain/blockdao/blockdao.go:113]: the current change is more efficient

Comment (dustinxie) [reply blockchain/blockdao/blockdao.go:113]: can actually remove fillWithBlockInfoAsTip, see 2nd commit

**Review by CoderZhi — APPROVED** [review]

File: blockchain/blockdao/blockdao.go [hunk 5]
```diff
@@ -163,12 +142,15 @@ func (dao *blockDAO) checkIndexers(ctx context.Context) error {
 			if producer == nil {
 				return errors.New("failed to get address")
 			}
-			ctxWithTip, err := dao.fillWithBlockInfoAsTip(ctx, bcCtx.ChainID, i-1)
-			if err != nil {
-				return err
+			if bcCtx.Tip.Height = tipBlk.Height(); bcCtx.Tip.Height > 0 {
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply blockchain/blockdao/blockdao.go:145]: same comment

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**